discovery-provider/src/tasks/index_spl_token.py

```diff
@@ -170,53 +173,87 @@ def parse_spl_token_transaction(
+        if version == 0 and meta.loaded_addresses:
+            writable = meta.loaded_addresses.writable or []
+            readonly = meta.loaded_addresses.readonly or []
+            account_keys.extend(str(key) for key in writable + readonly)
 
         memo_encoded = parse_memo_instruction(tx_message)
         vendor = decode_memo_and_extract_vendor(memo_encoded) if memo_encoded else None
 
-        sender_idx = get_account_index(instruction, SENDER_ACCOUNT_INDEX)
-        receiver_idx = get_account_index(instruction, RECEIVER_ACCOUNT_INDEX)
-        account_keys = list(map(lambda key: str(key), tx_message.account_keys))
-        sender_token_account = str(account_keys[sender_idx])
-        receiver_token_account = str(account_keys[receiver_idx])
-        sender_root_account = get_solana_tx_owner(meta, sender_idx)
-        receiver_root_account = get_solana_tx_owner(meta, receiver_idx)
-        balance_changes = get_solana_tx_token_balance_changes(
-            meta=meta, account_keys=account_keys
-        )
+        receiver_spl_tx_info: SplTokenTransactionInfo
 
-        # Balance is expected to change if there is a transfer instruction.
-        if receiver_token_account not in balance_changes:
-            logger.error(
-                f"index_spl_token.py | {tx_sig} error while parsing pre and post balances"
+        if version == 0:
+            balance_changes = get_solana_tx_token_balance_changes(
+                meta=meta, account_keys=account_keys, mint=WAUDIO_MINT
             )
-            return None
-        if balance_changes[receiver_token_account]["change"] == 0:
-            logger.error(f"index_spl_token.py | {tx_sig} no balance change found")
-            return None
-
-        receiver_spl_tx_info: SplTokenTransactionInfo = {
-            "user_bank": receiver_token_account,
-            "signature": tx_sig,
-            "slot": result.slot,
-            "timestamp": datetime.datetime.utcfromtimestamp(
-                float(result.block_time or 0)
-            ),
-            "vendor": vendor,
-            "prebalance": balance_changes[receiver_token_account]["pre_balance"],
-            "postbalance": balance_changes[receiver_token_account]["post_balance"],
-            "sender_wallet": sender_root_account,
-            "root_accounts": [sender_root_account, receiver_root_account],
-            "token_accounts": [sender_token_account, receiver_token_account],
-        }
-        return receiver_spl_tx_info
+            tx_infos: list[SplTokenTransactionInfo] = []
+            sender_token_accounts: list[str] = []
+            sender_root_accounts: list[str] = []
+            for key, balance_change in balance_changes.items():
+                if balance_change["change"] < 0:  # Only get senders
+                    sender_token_accounts.append(key)
+                    sender_root_accounts.append(balance_changes[key]["owner"])
+
+            for key, balance_change in balance_changes.items():
```

Comment:
just curious, are there actually multiple of these in practice?

Reply:
there are 3 (one with 0 change) in the example tx and we do need to iterate over them to find the non 0's. in practice though, it probably won't matter.

Reply:
I'm guessing 1 positive, 1 negative, so only one actual indexing thing happens?

Reply:
I guess we balance refresh external wallets with this change too, so that's valid

Reply:
yeah exactly^ we load up the negatives on the positives (the receiver)